src/main/java/com/microsoft/aad/msal4j/AuthenticationServiceException.java

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package com.microsoft.aad.msal4j;
+
+import lombok.Getter;
+import lombok.experimental.Accessors;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Exception type thrown when service returns an error response or other networking errors occur.
+ */
+@Accessors(fluent = true)
+@Getter
+public class AuthenticationServiceException extends AuthenticationException{
```

Comment:
>AuthenticationServiceException [](start = 13, length = 30)

Some of the errors should probably be categorized as client side / UiRequiredExceptions and not ServiceExceptions.

Reply:
Yes, namely "invalid_grant" service exceptions are exposed as UiRequiredServiceExceptions. Currently MSAL.NET treats all "invalid_grant" exceptions as UiRequiredExceptions, but I have a PR to filter some of these out (e.g. suberror "client_mismatch", which is a FOCI related suberror, is not a UiRequiredException).

Reply:
Consider having a factory that decides which type of exception needs to be thrown - ServiceException or UiRequiredException.

Reply:
@bgavrilMS I believe the service also returns "interaction_required". Are you treating these as UiRequiredExceptions as well?

Reply:
@sangonzal - yes, sorry, let me clarify. There are 2 cases where MSAL gets an error from the server and throws UiRequiredException: 

- Whenever it interacts with Evo and Evo responds with error code `invalid_grant`
- In the authorization flow, when the auth code url contains the param `error=login_required`

The MSAL library also throws `UiRequiredException` based on client-side logic, as follows: 

- in `AcquireTokenSilent` if the account / login_hint is null, or if a login_hint is used but no matching Account / too many Accounts are in the cache for that hint. 
- If no cache is configured 
- Some web UIs also throw this (should not apply to you, the system browser ui doesn't throw this)

Reply:
For the Authorization flow, that error will be sent when either: 

- Prompt.Never was passsed, but interaction is actually required
- An error occurs during silent auth that prevented the authentication to complete in time.